clients/wallet/src/stellar_wallet.rs

```diff
@@ -374,12 +378,112 @@ impl StellarWallet {
 			operations,
 		)?;
 
-		let _ = self.cache.save_tx_envelope(envelope.clone())?;
-
 		self.submit_transaction(envelope).await
 	}
 }
 
+pub enum Tx {
+	V0(TransactionV0),
+	V1(Transaction),
+}
+
+// Error handling methods for StellarWallet.
+impl StellarWallet {
+	pub async fn handle_error(&self, error: Error) -> Result<TransactionResponse, Error> {
+		match &error {
+			Error::HorizonSubmissionError { title: _, status: _, reason, envelope_xdr } => {
+				if let Some(env_xdr) = envelope_xdr {
+					let transaction_envelope =
+						TransactionEnvelope::from_base64_xdr(env_xdr).expect("should convert ok");
+					match &reason[..] {
+						"tx_bad_seq" =>
+							self.bump_sequence_number_and_submit(transaction_envelope).await,
```

Comment:
Here we would still need to add the check if the sequence number is greater or less than the accounts sequence number etc.

Reply:
> Right now, I've implemented the logic we've discussed on the Notion page: If the transaction sequence number is lower than the account sequence number, we discard the transaction. If the transaction sequence number is greater than the account sequence number, we wait until the transaction sequence number matches the correct one.

True, sorry I somehow missed that the logic here is kind of hidden in the preceding `!is_valid_seq_number` check. So it should indeed already work as proposed.

> Is it possible to have a sequence number collision? For example, if the account sequence number is 10, could two transactions with sequence number 14 exist? In that case, only one of them would be resubmitted.

In theory, yes. In practice, also yes but we try to prevent it with our [transaction_submission_lock](https://github.com/pendulum-chain/spacewalk/blob/7851e15a8f7c409347bd7808e6d05543589720c1/clients/wallet/src/stellar_wallet.rs#L42).
Or maybe this question has some more facets to it so let's dive a little deeper. 
If we consider the whole network level, ie. transactions created by possibly two or more different clients/wallets that happen to run in parallel: Let's say our account sequence number is 10, and we 'simultaneously' submit two different Stellar transactions from that account with sequence number 11. As both are valid, it's up to the Stellar horizon server to pick one of them and reject the other.
If we consider just our single vault client: It should never happen that we create two transactions with the same sequence number because we have this `transaction_submission_lock`. _But_ if we accidentally create two different instances of a `StellarWallet` that share the same secret key, then these instances have each their own `transaction_submission_lock` and this guarantee does not hold anymore. So we have to be careful with these instances.

> Instead of discarding the transaction, would it be a better idea to increase the sequence number when the transaction sequence number is lower than the account sequence number? We would not have any double-spending problem here (in theory) since once the transaction is submitted, it is removed from the cache. Am I right?"

If we did it this way, we'd risk double-payments because: Let's say our account sequence number is at 9 and we create a payment transaction with sequence number 10. Now we submit this transaction to the Stellar network with the `submit_transaction()` function but before reaching [this](https://github.com/pendulum-chain/spacewalk/blob/ad68c16f55141357d14df66d40717b42d089c96c/clients/wallet/src/stellar_wallet.rs#L84) line, our client executable suddenly dies for whatever reason. This means that we submitted the transaction to Stellar but it did not get removed from cache. Now, once the client restarts it will check the cache and see that this transaction is still contained. But when submitting it, it will receive a `tx_bad_seq` error because our account already reached a sequence number of 10. If we now naively decide to just bump the sequence number of this transaction to resubmit it, we do the same payment twice.

So...... maybe the only really foolproof solution here would be _not_ to discard the transaction but to first check the account's transactions on the network and check if this transaction was already executed or not. We can check the account's transactions and look for payments with exactly the same amount, asset and memo (since the payments our vault client does will _always_ have a memo attached to them).

Reply:
I'd suggest the following:
For transactions with an invalid (because *outdated*) sequence number, we query all Stellar transactions of that account and go through each one of them trying to find a `payment` operation that contains the same `TEXT_MEMO` as the invalid transaction we are dealing with. If we don't find any `payment` operation with the same memo, we bump the transaction's sequence number to an up-to-date one and re-submit.

WDYT @adelarja @b-yap?

Reply:
Sure.  Use [get_transactions_iter()](https://github.com/pendulum-chain/spacewalk/blob/ad68c16f55141357d14df66d40717b42d089c96c/clients/wallet/src/stellar_wallet.rs#L179) as inspo,  since it's a function that iterates BACKWARDS (from the newest to the oldest). There must be a filter though, until how far to look back.

Reply:
@ebma  @b-yap I like the idea. The responses memos has Option<Vec<u8>> type as we can see [here](https://github.com/pendulum-chain/spacewalk/blob/ad68c16f55141357d14df66d40717b42d089c96c/clients/wallet/src/horizon/responses.rs#L152). And the [Transaction](https://github.com/pendulum-chain/substrate-stellar-sdk/blob/master/src/xdr/types.rs#L3802) struct has Memo types. We can obtain the [MemoHash](https://github.com/pendulum-chain/substrate-stellar-sdk/blob/d1dd579cc27c614abe9eab7d0cdeae12ea735c0b/src/xdr/types.rs#L7217) from there, perform the necessary conversion, and compare, WDYT?

Reply:
Oh by the way, now that you mention it: We are not using `MemoHash` anymore but `MemoText` because unfortunately no Stellar wallet supports the `MemoHash` type somehow. To make it more convenient for users, we switched to using the `MemoText` instead (but are still storing some hashed data inside it).

Reply:
Great @ebma many thanks!